raft.go

```diff
@@ -935,6 +935,8 @@ func (r *raft) becomeLeader() {
 	// so the preceding log append does not count against the uncommitted log
 	// quota of the new leader. In other words, after the call to appendEntry,
 	// r.uncommittedSize is still 0.
+
+	r.raftLog.leaderTerm = r.Term // the leader's log is consistent with itself
```

Comment:
Unfortunately, the direct manipulation on `raftLog` and `Progress` fields is scattered across the `raft.go` too. This goes into the TODO above: all the correctness handling should be done at the level below.

In this particular case/PR though, we can move the `leaderTerm` update into the `r.appendEntry` call a few lines above, which in turn will delegate this to `r.raftLog.append` method. This will be more correct, I agree.

Reply:
Done this (see second commit).

Unfortunately, too many things have incorrect interfaces: the `raftLog` append and commit methods don't take the leader term into consideration. This lead to some tests testing scenarios which can never happen.

Maybe it would be good to reverse the order of commits here: first clean up the `raftLog` interfaces, and then this PR will be a small change.